aws_lambda_powertools/utilities/idempotency/persistence/redis.py

```diff
@@ -0,0 +1,440 @@
+        return self.expires_after_seconds
+
+    def _item_to_data_record(self, idempotency_key: str, item: Dict[str, Any]) -> DataRecord:
+        in_progress_expiry_timestamp = item.get(self.in_progress_expiry_attr)
+        if isinstance(in_progress_expiry_timestamp, str):
+            in_progress_expiry_timestamp = int(in_progress_expiry_timestamp)
+        return DataRecord(
+            idempotency_key=idempotency_key,
+            status=item[self.status_attr],
+            in_progress_expiry_timestamp=in_progress_expiry_timestamp,
+            response_data=str(item.get(self.data_attr)),
+            payload_hash=str(item.get(self.validation_key_attr)),
+            expiry_timestamp=item.get("expiration", None),
+        )
+
+    def _get_record(self, idempotency_key) -> DataRecord:
+        # See: https://redis.io/commands/get/
+        response = self.client.get(idempotency_key)
+        # key not found
+        if not response:
+            raise IdempotencyItemNotFoundError
+        try:
+            item = self._json_deserializer(response)
+        except json.JSONDecodeError:
+            raise IdempotencyOrphanRecordError
+        return self._item_to_data_record(idempotency_key, item)
+
+    def _put_in_progress_record(self, data_record: DataRecord) -> None:
+        item: Dict[str, Any] = {}
+        item = {
+            "name": data_record.idempotency_key,
+            "mapping": {
+                self.status_attr: data_record.status,
+                self.expiry_attr: data_record.expiry_timestamp,
+            },
+        }
+
+        if data_record.in_progress_expiry_timestamp is not None:
+            item["mapping"][self.in_progress_expiry_attr] = data_record.in_progress_expiry_timestamp
+
+        if self.payload_validation_enabled:
+            item["mapping"][self.validation_key_attr] = data_record.payload_hash
+
+        now = datetime.datetime.now()
+        try:
+            # |     LOCKED     |         RETRY if status = "INPROGRESS"                |     RETRY
+            # |----------------|-------------------------------------------------------|-------------> .... (time)
+            # |             Lambda                                              Idempotency Record
+            # |             Timeout                                                 Timeout
+            # |       (in_progress_expiry)                                          (expiry)
+
+            # Conditions to successfully save a record:
+
+            # The idempotency key does not exist:
+            #    - first time that this invocation key is used
+            #    - previous invocation with the same key was deleted due to TTL
+            #    - SET see https://redis.io/commands/set/
+
+            logger.debug(f"Putting record on Redis for idempotency key: {item['name']}")
+            encoded_item = self._json_serializer(item["mapping"])
+            ttl = self._get_expiry_second(expiry_timestamp=item["mapping"][self.expiry_attr])
+
```

Comment:
Suggestion: can we fetch the arguments from data_record instead of doing all this dictionary lookups? Seems we are wasting a lot of cycles here.

Reply:
yes we can use data_record instead